### ISSUE TITLE
False positive in derive_hash_xor_eq with multiple `Eq` implementations

### ISSUE THREAD

**Issue by llogiq** [rust-lang-nursery/rust-clippy#2025]
In [optional](https://github.com/llogiq/optional), I have both the standard derived `Eq` and `Hash`, plus one extra `Eq` impls on references for improved ergonomics. However, this additional `Eq' impl triggers the `derive_hash_xor_eq` lint. So perhaps the lint should look for existing derived implementations to avoid this false positive.

**Comment by oli-obk** [comment]
Can you create a minimal repro that we can use as a test case? Preferably without extra crates.

**Comment by llogiq** [comment]
https://play.rust-lang.org/?gist=18cdb2fd3b4b23e2d1d4d17c93537839&version=stable

```
use std::cmp::PartialEq;

#[derive(Debug, Eq, Hash, PartialEq)]
struct Foo;

impl<'a> PartialEq<Foo> for &'a Foo {
    fn eq(&self, _: &Foo) -> bool {
        true
    }
}

fn main() {
    assert_eq!(&Foo, Foo);
}
```

**Comment by vadixidav** [comment]
There also seems to be a false positive [here](https://github.com/rust-ndarray/ndarray/blob/2d7ce0789f0d0b789b08c4eaa6cee60f411a35b7/src/dimension/dim.rs#L73) in `ndarray`. It might be hard to track down why this is happening. I am just reporting it here.

**Comment by lo48576** [comment]
Now it seems that `derive_ord_xor_partial_ord` (#1621, #5848) has the same false positive.

https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=12f17e27712b221f20555aab95a51869

```
    Checking playground v0.0.1 (/playground)
error: you are deriving `Ord` but have implemented `PartialOrd` explicitly
  --> src/lib.rs:4:37
   |
4  | #[derive(PartialEq, Eq, PartialOrd, Ord, Hash)]
   |                                     ^^^
   |
   = note: `#[deny(clippy::derive_ord_xor_partial_ord)]` on by default
note: `PartialOrd` implemented here
  --> src/lib.rs:14:1
   |
14 | / impl PartialOrd<Foo> for &Foo {
15 | |     fn partial_cmp(&self, o: &Foo) -> Option<std::cmp::Ordering> {
16 | |         // Redirect to the derived PartialOrd implementation.
17 | |         (**self).partial_cmp(o)
18 | |     }
19 | | }
   | |_^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#derive_ord_xor_partial_ord
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error

error: could not compile `playground`

To learn more, run the command again with --verbose.
```

**Comment by Jarcho** [comment]
This has already been fixed.
